### PR TITLE
SALTO-3836 Add post deploy action for app

### PR DESCRIPTION
**Pull request by idoamit8**
1. Added a post deploy action for additional apps to ensure integration with Okta and service provider.  
2. Added test to check the new validator. 

---

_Additional context for reviewer_
1. This PR is rebased on another PR because the dependency of client.baseUrl 
1. Please check only the last commit named 'add serrvice_provider_connection"
---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import { ChangeError, ChangeValidator, InstanceElement, getChangeData, isAdditionChange, isInstanceChange } from '@salto-io/adapter-api'
+import OktaClient, { getAdminUrl } from '../client/client'
+import { APPLICATION_TYPE_NAME } from '../constants'
+
+const SAML_2_0_APP = 'SAML_2_0'
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:21]: nit: We already use this string in `app_deployment.ts` so I think it's best to move to `constants.ts` and reuse :)

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          'Click on "Sign On" tab.',
+          'On the right side, click on "View SAML setup instructions".',
+          'Follow the instructions to complete the setup.',
+        ],
+      },
+    },
+  }
+}
+
+
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:47]: nit: extra linebreak 🙃

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+* limitations under the License.
+*/
+
+import { ChangeError, ChangeValidator, InstanceElement, getChangeData, isAdditionChange, isInstanceChange } from '@salto-io/adapter-api'
+import OktaClient, { getAdminUrl } from '../client/client'
+import { APPLICATION_TYPE_NAME } from '../constants'
+
+const SAML_2_0_APP = 'SAML_2_0'
+
+const createChangeError = (instance: InstanceElement, baseUrl: string): ChangeError => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:23]: I think we can have a more intuitive name here, WDYT?

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  const suffixUrl = '/admin/apps/active'
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:24]: Is this the url when `Application` is in status `INACTIVE` as well?

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    severity: 'Info',
+    message: 'New app integration setup required',
+    detailedMessage: 'In the service provider, follow the instructions provided by Okta to configure the app integration',
+    deployActions: {
+      postAction: {
+        title: 'New app integration setup required',
+        description: 'To complete the setup of the new app integration in Okta, follow these steps:',
+        subActions: [
+          `Go to application page at ${url}`,
+          `Click on ${instance.value.name ?? instance.value.customName ?? 'the application'}.`,
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:37]: I think we should use `instance.value.label` here, because:
1. That's the name visible to the user in Okta's UI
2. `customName` is hidden so we won't have it in addition changes

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+const SAML_2_0_APP = 'SAML_2_0'
+
+const createChangeError = (instance: InstanceElement, baseUrl: string): ChangeError => {
+  const suffixUrl = '/admin/apps/active'
+  const url = (new URL(suffixUrl, baseUrl)).href
+  return {
+    elemID: instance.elemID,
+    severity: 'Info',
+    message: 'New app integration setup required',
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:29]: nit: maybe application instead of app?

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+          'On the right side, click on "View SAML setup instructions".',
+          'Follow the instructions to complete the setup.',
+        ],
+      },
+    },
+  }
+}
+
+
+export const serviceProviderConnectionValidator: (client: OktaClient) =>
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/service_provider_connection.ts:48]: Please add a docstring explaining what this CV is for

**Review by shir-reifenberg — COMMENTED** [review]
Looks great! :) Added a few comments & questions, but nothing major.
Also, I think file name for `service_provider_connection` should be more indicative (e.g `app_integration_setup`, `application_addition_setup` or anything similar)

File: packages/okta-adapter/src/change_validators/service_provider_connection.ts (hunk 4 above)
Comment (idoamit8) [reply packages/okta-adapter/src/change_validators/service_provider_connection.ts:24]: As I can see in the website than yes. This is the URL also for inactive apps

File: packages/okta-adapter/src/change_validators/app_integration_setup.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  const url = baseUrl ? (new URL(suffixUrl, baseUrl)).href : undefined
+  return {
+    elemID: instance.elemID,
+    severity: 'Info',
+    message: 'New application integration setup required',
+    detailedMessage: 'In the service provider, follow the instructions provided by Okta to configure the app integration',
+    deployActions: {
+      postAction: {
+        title: 'New application integration setup required',
+        description: 'To complete the setup of the new app integration in Okta, follow these steps:',
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/app_integration_setup.ts:32]:
```suggestion
        description: 'To complete the setup of the new application integration in Okta, follow these steps:',
```

File: packages/okta-adapter/test/change_validators/app_integration_setup.test.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            `Click on ${appInstance.value.name}.`,
+            'Click on "Sign On" tab.',
+            'On the right side, click on "View SAML setup instructions".',
+            'Follow the instructions to complete the setup.',
+          ],
+        },
+      },
+    }])
+  })
+  it('should return empty array if app has not Saml sign on mode', async () => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/test/change_validators/app_integration_setup.test.ts:71]: nit
```suggestion
  it('should return empty array if app has no Saml sign on mode', async () => {
```

File: packages/okta-adapter/src/change_validators/app_integration_setup.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+          'Click on "Sign On" tab.',
+          'On the right side, click on "View SAML setup instructions".',
+          'Follow the instructions to complete the setup.',
+        ],
+      },
+    },
+  }
+}
+/**
+ * Ensures that a service provider application integrates with Okta using the SAML 2.0 protocol.
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/app_integration_setup.ts:45]: I think this docstring is not that clear... we should explain what the CV is doing and why it is needed, so i'd rephrase to something like: `Integrating SAML application requires additional setup in the service provider. This validator provides instructions for the user on how to complete this setup`

**Review by shir-reifenberg — APPROVED** [review]
💯 👏 
Added only small comments. Please merge only after https://github.com/salto-io/salto/pull/4220 approved & merged 🙏 
Also, seems like coverage decreased, might be related to the prev PR as well but worth checking
